src/keboola_mcp_server/sql_tools.py

```
@@ -24,26 +156,22 @@ async def query_table(
     * SQL queries must include the fully qualified table names including the database name, e.g.:
       SELECT * FROM "db_name"."db_schema_name"."table_name";
     """
-    connection_manager = ctx.session.state["connection_manager"]
-    assert isinstance(connection_manager, ConnectionManager)
-
-    with connection_manager.create_snowflake_connection() as conn:
-        try:
-            cursor = conn.cursor()
-            cursor.execute(sql_query)
-            result = cursor.fetchall()
-            columns = [col[0] for col in cursor.description]
-
-            # Convert to CSV
-            output = StringIO()
-            writer = csv.writer(output)
-            writer.writerow(columns)
-            writer.writerows(result)
+    workspace_manager = WorkspaceManager.from_state(ctx.session.state)
+    result = await workspace_manager.execute_query(sql_query)
+    if result.is_ok:
+        if result.data:
+            data = result.data
+        else:
+            # non-SELECT query, this should not really happen, because this tool is for running SELECT queries
+            data = SqlSelectData(columns=["message"], rows=[{"message": result.message}])
```

When it should not happen. Should we raise an exception or is there a reason we need to handle this case?

---

The query tools is called by the LLM and it is the LLM which generates the SQL passed to the tool. The tool explicitly says that it expects the SQL to be a SELECT statement, but the actual SQL statement passed to the tool by the LLM can be anything. IMO it still makes sense to try to run that SQL (even if it is not a SELECT statement) and pass its data back to the LLM.

---

Sure, you are right.